### PR TITLE
Implement loading namespaces with the `using` keyword

### PR DESCRIPTION
**Pull request by ghallak**
Fixes: #233 
Fixes: #234

### CODE REVIEW

**Review by radrow — APPROVED** [review]
Super!

File: src/aeso_ast_infer_types.erl [hunk 1]
```diff
@@ -312,9 +317,48 @@ bind_contract({Contract, Ann, Id, Contents}, Env)
...
+        Namespaces ->
+            IsVisible = fun(Namespace) ->
+                            case Namespace of
+                                {_, _, {for, Names}} ->
+                                    case lists:member(Name, Names) of
+                                        true ->
+                                            true;
+                                        false ->
+                                            false
+                                    end;
```
Comment (radrow) [inline src/aeso_ast_infer_types.erl:348]:
```suggestion
                                    lists:member(Name, Names)
```

File: src/aeso_ast_infer_types.erl [hunk 2]
```diff
@@ -312,9 +317,48 @@ bind_contract({Contract, Ann, Id, Contents}, Env)
...
+                                {_, _, {hiding, Names}} ->
+                                    case lists:member(Name, Names) of
+                                        true ->
+                                            false;
+                                        false ->
+                                            true
+                                    end;
```
Comment (radrow) [inline src/aeso_ast_infer_types.erl:355]:
```suggestion
                                    not lists:member(Name, Names)
```

File: src/aeso_ast_infer_types.erl (hunk 1 above)
Comment (ghallak) [reply src/aeso_ast_infer_types.erl:348]: I'm surprised that I wrote it this way. Will change.

**Review by radrow — APPROVED** [review]

**Comment by hanssv** [comment]
Sorry I'm late to the party, but I was thinking about the example `using_namespace_ambiguous_name` - isn't the problem that you alias both `Xa` and `Xb` to `A` and the error is really on the line `using Xb as A` - the example reports two (other) errors? Otherwise it looks good to me.

Also there is a trivial conflict in documentation it seems.

**Comment by ghallak** [comment]
> isn't the problem that you alias both `Xa` and `Xb` to `A` and the error is really on the line `using Xb as A` - the example reports two (other) errors?

@hanssv I would say that it's not an error that two namespaces have the same alias, as these namespaces might have functions with different names:

```
namespace Xa = function f() = 1
namespace Xb = function g() = 2
contract Cntr =
  using Xa as A
  using Xb as A
  type state = int
  entrypoint init() = A.f() + A.g()
```

I think the above example should compile, and I can't think of a way that this could cause any bugs (beside that, haskell allows using the same alias for multiple modules).

**Comment by hanssv** [comment]
Again, let me remind me about the target audience... Modeling things after Haskell might or might not be the correct thing 😅 

That last example did not match my mental picture of things but I should be able to adapt the picture - possibly the documentation could clarify this?

**Review by hanssv — DISMISSED** [review]

**Comment by ghallak** [comment]
@hanssv I've updated the docs to here 7fabbff11e07da8451df3d57a281f9bae7b3d739

**Review by hanssv — APPROVED** [review]
👍

**Pull request merged**